src/r1cs/mod.rs

```diff
@@ -159,6 +160,76 @@ impl<E: Engine> R1CSShape<E> {
     })
   }
 
+  /// Generate a random [R1CSShape] with the specified number of constraints, variables, and public inputs/outputs.
+  pub fn random<R: RngCore + CryptoRng>(
```

Comment:
Just a thought, if we're going to implement `random` recursively on some structs, could we possibly use `Arbitrary` from [proptest](https://github.com/proptest-rs/proptest)? We'd get randomized testing as a side-effect.

Reply:
I'm having trouble finding documentation on how to do it, but I'll create an issue to tackle it later.